src/EditModel/event-program/epics.js

```diff
@@ -83,132 +148,311 @@ function loadEventProgramMetadataByProgramId(programId) {
+*  Copies the original model, that is used if the user cancels editing of the model */
+export const editTrackerProgramStage = action$ =>
+    action$
+        .ofType(PROGRAM_STAGE_EDIT)
+        .map(action => action.payload)
+        .flatMap(({ stageId }) =>
+            eventProgramStore
+                .take(1)
+                .map(get('programStages'))
+                .map(programStages => {
+                    const index = programStages.findIndex(
+                        stage => stage.id == stageId
+                    );
+                    const model = programStages[index].clone();
+                    const setter = { programStageToEditCopy: model };
+
+                    eventProgramStore.setState(setter);
+                })
+        )
+        .flatMapTo(Observable.of({ type: 'EMPTY' }));
+
+export const saveTrackerProgramStage = action$ =>
+    action$
+        .ofType(PROGRAM_STAGE_EDIT_SAVE)
+        .flatMap(action =>
+            eventProgramStore.take(1).map(store => {
+                const stageId = store.programStageToEditCopy.id;
+                const index = store.programStages.findIndex(
+                    stage => stage.id == stageId
+                );
+                if (index < 0) {
+                    console.warn(
+                        `ProgramStage with id ${stageId} does not exist`
+                    );
+                }
+                try {
+                    eventProgramStore.setState({
+                        programStageToEditCopy: null,
+                    });
+                } catch (e) {
+                    console.log(e);
+                }
             })
-    );
+        )
+        .flatMapTo(Observable.of({ type: PROGRAM_STAGE_EDIT_RESET }));
+
+export const cancelProgramStageEdit = action$ =>
+    action$
+        .ofType(PROGRAM_STAGE_EDIT_CANCEL)
+        .flatMap(() =>
+            eventProgramStore.take(1).map(store => {
+                const stageId = store.programStageToEditCopy.id;
+                const index = store.programStages.findIndex(
+                    stage => stage.id == stageId
+                );
+                if (index < 0) {
+                    console.warn(
+                        `ProgramStage with id ${stageId} does not exist`
+                    );
+                }
+                const model = store.programStageToEditCopy;
+                try {
+                    eventProgramStore.setState(
+                        set(`programStages[${index}]`)(model, {
+                            ...eventProgramStore.getState(),
+                        }),
+                        set(
+                            'programStageToEditCopy',
+                            null,
+                            eventProgramStore.getState()
+                        )
+                    );
+                } catch (e) {
+                    console.log(e);
+                }
+            })
+        )
```

Comment:
This nested function from L413-440 could be moved into a named function and the logic itself could be split into smaller functions as there are multiple things happening. One function could look for the id, and pass it to the next which does the state operations.

I also find it a bit strange to first log out that the id doesn't exist on L420-422, and then use that non-existing id to set something. Wouldn't that operation fail and log out the error on L437?